content/post/20220520-quantum-compile.md

```diff
@@ -0,0 +1,471 @@
+
+```lisp
+  (define-type AlgNum
+    "Represents the algebraic number a1 + a2 * sqrt(2)."
+    (AlgNum Integer Integer))
+
+  (define-instance (Eq AlgNum)
+    (define (== a b)
+      (match (Tuple a b)
+        ((Tuple (AlgNum a1 a2) (AlgNum b1 b2))
+         (and (== a1 b1) (== a2 b2))))))
+  
+  (define-instance (Num AlgNum)
+    (define (+ a b)
+      (match (Tuple a b)
+        ((Tuple (AlgNum a1 a2) (AlgNum b1 b2))
+         (AlgInt (+ a1 b1) (+ a2 b2)))))
+    (define (- a b)
+      (match (Tuple a b)
+        ((Tuple (AlgNum a1 a2) (AlgNum b1 b2))
+         (AlgInt (- a1 b1) (- a2 b2)))))
+    (define (* a b)
+      (match (Tuple a b)
+        ((Tuple (AlgNum a1 a2) (AlgNum b1 b2))
+         (AlgInt (+ (* a1 b1) (* 2 (* a2 b2)))
+                 (+ (* a1 b2) (* a2 b1))))))
+    (define (fromInt n)
+      (AlgNum n 0)))
+```
+
+We can verify it works by seeing that $(1-2\sqrt{2})^2 = 9-4\sqrt{2}$:
+
+```lisp
+> (coalton (* (AlgNum 1 -2) (AlgNum 1 -2)))
+#.(ALGNUM 9 -4)
+```
+
+Since `Num` requires `fromInt` to be defined, any `Num` type, including our own `AlgNum`, permits overloading integer syntax. We don't have to laboriously write `(fromInt 2)` every time we want to use integer syntax as a shorthand for our `AlgNum` type.
+
+```lisp
+> (coalton (* 2 (AlgNum 0 -1)))
+#.(ALGNUM 0 -2)
+```
+
+As is to be expected at this point, functions on `AlgInt` are inferred appropriately.
+
+```lisp
+(define (algebraic-conjugate x)
+  (match x
+    ((AlgNum a b) (AlgNum a (negate b)))))
+```
+
+We can see that
+
+```lisp
+> (type-of 'algebraic-conjugate)
+(ALGNUM → ALGNUM)
+```
+
+Modeling these algebras works out quite well, especially when we have more of them, matrices of them, vectors of them, polynomials of them, and so on.
+
+Though perhaps not as effective as mathematics itself, Coalton's facilities handle the Selinger  Clifford+T decomposition algorithm handsomely. This lead to the introduction of a first-of-its-kind feature to a general-purpose quantum compiler: the ability to do discrete compilation of arbitrary quantum programs. This is now implemented as a part of the [`cl-quil/discrete` contrib module](https://github.com/quil-lang/quilc/tree/master/src/discrete) to quilc.
+
+We can repeat the compilation of the `CZ` gate that we did before for the IBM chip instead for a made-up chip of two qubits that only supports the Clifford+T set.
+
+```
+$ echo 'PRAGMA TOLERANCE "0.001"; CZ 0 1' | ./quilc --isa discrete2
+S 0
+S 0
+H 1
+CNOT 0 1
+H 0
+T 0
+S 0
+H 0
+S 0
+S 0
+H 0
+T 0
+S 0
+H 0
+H 1
+```
+
+Notice how every operation in the output is either `H`, `S`, `T`, or `CNOT`. There are only two differences in how we run the compiler between this and the previous IBM-chip compilation:
+
+1. We had to specify the chip `discrete2`. (Quilc will be happy to construct a chip of any requested topology; this is just a convenient built-in.)
+2. We had to specify a tolerance, here a discrete compilation accuracy within 0.1%.
+
```

Comment:
Add the output from QVM after running the discrete and continuous gate sets

Reply:
added todo...